Mathlib/Analysis/InnerProductSpace/JointEigenspace.lean

```diff
@@ -63,50 +65,127 @@ theorem orthogonalFamily_eigenspace_inf_eigenspace (hA : A.IsSymmetric) (hB : B.
+  conv_lhs => rw [← Submodule.map_iSup]
+  conv_rhs => rw [← map_subtype_top F]
+  congr!
+  have H : IsSymmetric (S.restrict hInv) := fun x y ↦ hS (F.subtype x) y
+  apply orthogonal_eq_bot_iff.mp (H.orthogonalComplement_iSup_eigenspaces_eq_bot)
+
+/-- In finite dimensions, the indexed supremum of joint `maxGenEigenspaces` of a finite tuple of
+commuting operators equals `⊤` provided the indexed supremum of `maxGenEigenspaces` of each
+operator in the tuple equals `⊤`. -/
+theorem iSup_iInf_maxGenEigenspace_eq_top_of_iSup_maxGenEigenspace_eq_top_of_commute {ι K V : Type*}
+    [Field K] [AddCommGroup V] [Module K V] [FiniteDimensional K V]
+    (f : ι → Module.End K V) (h : Pairwise fun i j ↦ Commute (f i) (f j))
+    (h' : ∀ i, ⨆ μ, (f i).maxGenEigenspace μ = ⊤) :
+    ⨆ χ : ι → K, ⨅ i, (f i).maxGenEigenspace (χ i) = ⊤ := by
+  refine Module.End.iSup_iInf_maxGenEigenspace_eq_top_of_forall_mapsTo _
+    (fun i j ↦ Module.End.mapsTo_maxGenEigenspace_of_comm ?_) h'
+  rcases eq_or_ne j i with rfl | hij <;> tauto
+
+/-- Every generalized eigenspace of a symmetric operator is an eigenspace. -/
+theorem genEigenspace_eq_eigenspace
+    {T : E →ₗ[𝕜] E} (hT : T.IsSymmetric) {n : ℕ} {μ : 𝕜} (hn : 1 ≤ n) :
+    genEigenspace T μ n = genEigenspace T μ 1 := by
+  refine Nat.le_induction rfl (fun k hk ih ↦ ?_) n hn
+  refine ih ▸ le_antisymm (fun x hx ↦ ?_) ((genEigenspace T μ).mono k.le_succ)
+  obtain (rfl | hx_ne) := eq_or_ne x 0
+  · exact zero_mem _
+  · have hμ : HasEigenvalue T μ := hasEigenvalue_of_hasGenEigenvalue (k := k + 1) <|
+      (genEigenspace T μ (k + 1)).ne_bot_iff.mpr ⟨x, hx, hx_ne⟩
+    have hT' := LinearMap.isSymmetric_iff_isSelfAdjoint T |>.mp hT
+    have hTμ : ((T - μ • 1) ^ k).IsSymmetric  := by
+      rw [LinearMap.isSymmetric_iff_isSelfAdjoint]
+      refine .pow (hT'.sub (.smul ?_ ?_)) k
+      · exact hT.conj_eigenvalue_eq_self hμ
+      · exact (LinearMap.isSymmetric_iff_isSelfAdjoint 1).mp LinearMap.IsSymmetric.id
+    rw [mem_genEigenspace, ← norm_eq_zero, ← sq_eq_zero_iff, norm_sq_eq_inner (𝕜 := 𝕜)]
+    rw [hTμ, ← LinearMap.comp_apply, ← LinearMap.mul_eq_comp, ← pow_add]
+    simp [mem_genEigenspace .. |>.mp <| (genEigenspace T μ).mono (show k + 1 ≤ k + k by gcongr) hx]
+
+lemma maxGenEigenspace_eq_eigenspace
+    {T : E →ₗ[𝕜] E} (hT : T.IsSymmetric) {μ : 𝕜} :
+    maxGenEigenspace T μ = eigenspace T μ := calc
+  _ = ⨆ n, genEigenspace T μ (n + 1) := by
+    rw [maxGenEigenspace_def, ← sup_iSup_nat_succ, genEigenspace_def]; simp [LinearMap.one_eq_id]
+  _ = ⨆ _ : ℕ, genEigenspace T μ 1 := by
+    congr! 2 with n; exact genEigenspace_eq_eigenspace hT n.succ_pos
+  _ = eigenspace T μ := by simp [genEigenspace_def, eigenspace_def]
+
+/-- In finite dimensions, the indexed supremum of the joint eigenspaces of a commuting tuple of
+symmetric linear operators equals `⊤`. -/
+theorem iSup_iInf_eq_top_of_commute {ι : Type*}
+    (T : ι → E →ₗ[𝕜] E) (hT : ∀ i, (T i).IsSymmetric) (h : Pairwise fun i j ↦ Commute (T i) (T j)):
```

Comment:
I'm still not convinced `T` should be explicit here (or anywhere else)
```suggestion
theorem iSup_iInf_eq_top_of_commute {ι : Type*} (T : ι → E →ₗ[𝕜] E)
    (hT : ∀ i, (T i).IsSymmetric) (h : Pairwise fun i j ↦ Commute (T i) (T j)) :
```

Reply:
Let's make it implicit, then! Worst case, it will be changed back. It can surely be inferred by the `hT`...